dask/dataframe/io/parquet/arrow.py

```diff
@@ -164,12 +166,22 @@ def _get_all_partition_keys(ds):
     """
     categories = defaultdict(list)
     pkeys = defaultdict(list)
+    file_frag = None
     for file_frag in ds.get_fragments():
         keys = pa_ds._get_partition_keys(file_frag.partition_expression)
         pkeys[file_frag.path] = keys
         for k, v in keys.items():
             if v not in categories[k]:
                 categories[k].append(v)
+
+    if file_frag is not None:
+        # Check/correct order of `categories` using last file_frag
+        cat_keys = [
+            o.split(" =")[0]
+            for o in re.findall("[^(]* =", str(file_frag.partition_expression))
```

Comment:
I'd love to see something like this live in upstream

Reply:
Yes, this is exactly my sentiment.  Unfortunately, the exposed API and partitioning information is far from consistent between the three supported engines.  This inconsistency is the reason I originally went with the general regex approach for specifying an aggregation range.

At this point, I am sort of throwing up my hands.  It seems like we need to do something messy like this to produce a reasonable API.  Other ideas are welcome :)

Reply:
Well, this affects only arrow, doesn't it? I agree that arrow might not want to expose the exact signature we're requiring here but IIUC this function is something like an aggregation of filepath encoded data.

```
A=1/B=2/data.parquet
A=4/B=2/data.parquet
```

I am struggling a bit to parse what the expected output would be for that simple example but I expect something like

```python
{
    "A=1/B=2/data.parquet": ["A", "B"],
    "A=4/B=2/data.parquet": ["A", "B"],
}
```

or

```python
{
    "A=1/B=2/data.parquet": [1, 2],
    "A=4/B=2/data.parquet": [4, 2],
}
```

Maybe a short example in the doc string would be helpful. imho, nomenclature is not sufficiently aligned and I always get confused when reading about 'categories', 'fragments', 'partitions', etc.

Reply:
>Well, this affects only arrow, doesn't it?

Yes, thats right - I was conflating this specific change with a larger challenge I was facing :)

I agree that this specific logic should live in pyarrow.  There is currently a (private) `_get_partition_keys` utility to convert the "partition expression" into a list of "keys".  However, the result is **not** ordered in the way we need it to be, hence this custom/ugly logic.  @jorisvandenbossche may be able to comment, but I suspect I should just raise a jira/pyarrow issue about it.

Reply:
We still have open https://issues.apache.org/jira/browse/ARROW-8655 to better expose the partitioning information. Concrete suggestions on what information you want to have accessible in which way is always welcome.

> However, the result is **not** ordered in the way we need it to be, hence this custom/ugly logic

Ah, that's useful feedback, that might be an easy change in pyarrow (the implementation in C++ is currently using an unordered map, I notice. I suppose we can change that to something that preserves the order).

---

About the specific code line this is commenting on: I would suggest to _not_ rely on the `repr` of the partition expression (that seems rather brittle, and I can't guarantee any stability for the _display_ of an expression). As an alternative short-term work-around, can you do something similar on the file path?

Reply:
note that this regex does not allow `(` in the category names. For "proper" dataset formats I would expect this to not be an issue since the paths are likely encoded, e.g. url encoded, and `(` _might_ be forbidden but it could pop up and this regex would return nonsense.

Reply:
>About the specific code line this is commenting on: I would suggest to not rely on the repr of the partition expression (that seems rather brittle, and I can't guarantee any stability for the display of an expression). As an alternative short-term work-around, can you do something similar on the file path?

I agree that it is not ideal to depend on the `repr`.  We can use the file path for now, but that is also a bit tricky.  Does anyone know if the following would work for both linux and windows?...

```python
        cat_keys = [
            o.split("=")[0]
            for o in re.findall(f"[^{fs.sep}]*=", str(file_frag.partition_expression))
        ]
```

Reply:
fsspec has a `make_posix_path` which normalizes `/` or `\`, but not sure if that's a public function. 

But I also _think_ that `file_frag.path` will always return normalized paths (using `/`) even on Windows. So hardcoding `"[/]*="` in the regex could also work.

Reply:
Yes it's public, but "/" is the only variant that gets used in practice. Note that "/" paths always work on any OS (python has some in-built hidden windows stuff to do this, and we can rely on it).

Reply:
I don't see why it shouldn't. Tests cover windows, if you are sure this path is tested a green build should be sufficient to verify ;)

Reply:
if in doubt, you can also use smth like `os.path.split` which should resolve properly for win/linux but is likely slower